File: android/app/src/androidTest/java/host/exp/exponent/utils/TestReporterRule.java
// Copyright 2015-present 650 Industries. All rights reserved.

package host.exp.exponent.utils;

import android.util.Log;

import org.junit.rules.TestRule;
import org.junit.runner.Description;
import org.junit.runners.model.Statement;

public class TestReporterRule implements TestRule {

  private static final String TAG = "EXPO_TEST_REPORTER";

  private String mLogs = "";

  public void logTestInfo(final String log) {
    mLogs += log + "\n";
  }

  private void reportResult(final boolean success, final String testName) {
    try {
      TestServerUtils.reportTestResult(success, testName, mLogs);
    } catch (Exception e) {
      e.printStackTrace();
    }
  }

  @Override
  public Statement apply(Statement base, Description description) {
    return statement(base, description);
  }

  private Statement statement(final Statement base, final Description description) {
    return new Statement() {
      @Override
      public void evaluate() throws Throwable {
        Log.d(TAG, "start_test_" + description.getDisplayName());
        try {
          mLogs = "";
          base.evaluate();
          Log.d(TAG, "end_test_" + description.getDisplayName());
          reportResult(true, description.getDisplayName());
          return;
        } catch (Throwable t) {
          Log.d(TAG, "end_test_" + description.getDisplayName());
          reportResult(false, description.getDisplayName());
          throw t;
        }
      }
    };
  }
}

File: android/app/src/androidTest/java/host/exp/exponent/utils/TestReporterRule.kt
// Copyright 2015-present 650 Industries. All rights reserved.
package host.exp.exponent.utils

import android.util.Log
import org.junit.rules.TestRule
import org.junit.runner.Description
import org.junit.runners.model.Statement
import java.lang.Exception

class TestReporterRule : TestRule {
  private var logs = ""
  fun logTestInfo(log: String) {
    logs += log + "\n"
  }

  private fun reportResult(success: Boolean, testName: String) {
    try {
      TestServerUtils.reportTestResult(success, testName, logs)
    } catch (e: Exception) {
      e.printStackTrace()
    }
  }

  override fun apply(base: Statement, description: Description): Statement {
    return statement(base, description)
  }

  private fun statement(base: Statement, description: Description): Statement {
    return object : Statement() {
      @Throws(Throwable::class)
      override fun evaluate() {
        Log.d(TAG, "start_test_" + description.displayName)
        try {
          logs = ""
          base.evaluate()
          Log.d(TAG, "end_test_" + description.displayName)
          reportResult(true, description.displayName)
          return
        } catch (t: Throwable) {
          Log.d(TAG, "end_test_" + description.displayName)
          reportResult(false, description.displayName)
          throw t
        }
      }
    }
  }

  companion object {
    private const val TAG = "EXPO_TEST_REPORTER"
  }
}
